src/Checkpointer.jl

```diff
@@ -78,18 +96,188 @@ end
+            auto_restart_model_cache!(sim, comms_ctx, checkpoint_dir)
+        end
+    end
+    auto_restart_coupler_fields!(cs, checkpoint_dir)
+end
+
+"""
+    restart!(cs::CoupledSimulation, checkpoint_dir, checkpoint_t)
+
+Overwrite the content of `cs` with checkpoints in `checkpoint_dir` at time `checkpoint_t`.
+
+Return a true if the simulation was restarted.
+"""
+function restart!(cs, checkpoint_dir, checkpoint_t)
+    pid = ClimaComms.mypid(cs.comms_ctx)
+    for sim in cs.model_sims
+        if Checkpointer.get_model_prog_state(sim) !== nothing
+            input_file_state =
+                output_file = joinpath(checkpoint_dir, "checkpoint_$(Interfacer.name(sim))_$(checkpoint_t).hdf5")
+            restart_model_state!(sim, input_file_state, cs.comms_ctx)
+            input_file_cache =
+                joinpath(checkpoint_dir, "checkpoint_cache_$(pid)_$(Interfacer.name(sim))_$(checkpoint_t).jld2")
+            restart_model_cache!(sim, input_file_cache)
         end
     end
+    input_file_coupler_fields = joinpath(checkpoint_dir, "checkpoint_coupler_fields_$(pid)_$(checkpoint_t).jld2")
+    restart_coupler_fields!(cs, input_file_coupler_fields)
+    return true
+end
+
+"""
+    auto_restart_coupler_fields!(cs, checkpoint_dir)
+
+Find checkpoints in `checkpoint_dir` for the coupler fields and write them in `cs`.
+"""
+function auto_restart_coupler_fields!(cs, checkpoint_dir)
+    isdir(checkpoint_dir) || return false
+    pid = ClimaComms.mypid(cs.comms_ctx)
+    restart_file_rx = Regex("checkpoint_coupler_fields_$(pid)_\\d+.jld2")
+    restarts = filter(f -> !isnothing(match(restart_file_rx, f)), readdir(checkpoint_dir))
+    isempty(restarts) && return false
+    input_file = joinpath(checkpoint_dir, last(sort_by_creation_time(restarts)))
+    @info "Restoring coupler fields from $(input_file)"
+    restart_coupler_fields!(cs, input_file)
+    return true
 end
 
+"""
+    restart_model_cache!(sim::Interfacer.ComponentModelSimulation, comms_ctx::ClimaComms.AbstractCommsContext, t::Int; input_dir = "input")
+
+Sets the model cache of a simulation from a collection of JLD2 files from a given time, t (in seconds).
+"""
+function auto_restart_model_cache!(
+    sim::Interfacer.ComponentModelSimulation,
+    comms_ctx::ClimaComms.AbstractCommsContext,
+    checkpoint_dir,
+)
+    isdir(checkpoint_dir) || return false
+    pid = ClimaComms.mypid(comms_ctx)
+    restart_file_rx = Regex("checkpoint_cache_$(pid)_$(Interfacer.name(sim))_\\d+.jld2")
+    restarts = filter(f -> !isnothing(match(restart_file_rx, f)), readdir(checkpoint_dir))
+    isempty(restarts) && return false
+    input_file = joinpath(checkpoint_dir, last(sort_by_creation_time(restarts)))
+    @info "Restarting $(Interfacer.name(sim)) cache from $input_file"
+    return true
+end
+
+function restart_model_cache!(sim, input_file)
+    ispath(input_file) || error("File $(input_file) not found")
+    restore_cache!(sim, JLD2.jldopen(input_file)["cache"])
```

Comment:
It would be helpful to note here again that `restore_cache!` will be implemented by component models with caches, so readers know where to look for the implementation

Reply:
Added in docstring and comment